tt_metal/api/program.hpp

```diff
@@ -0,0 +1,162 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <string_view>
+#include "types.hpp"
+
+#include "tt_metal/impl/kernels/kernel_types.hpp"
+#include "tt_metal/impl/buffers/circular_buffer_types.hpp"
+
+//==================================================
+//                  PROGRAM MANAGEMENT
+//==================================================
+
+namespace tt::tt_metal{
+namespace v1 {
+
+/**
+ * @brief Creates a Program object, which bundles kernels, circular buffers, and semaphores for execution on the device.
+ *
+ * @return Program handle to the created program.
+ */
+Program CreateProgram();
+
+
+/**
+ * @brief Creates a data movement or compute kernel and adds it to the program.
+ *
+ * @param program The program to which this kernel will be added.
+ * @param file_name Path to the kernel source file.
+ * @param core_spec Specifies the cores on which the kernel will be placed.
+ * @param config DataMovementConfig for the kernel.
+ * @return KernelHandle representing the kernel ID.
+ */
+KernelHandle CreateKernel(
```

Comment:
🧵  Different thread related to error handling. General design guidance - prevent or catch errors early. 

Do you see ways to prevent errors described in the above message?

In approach with object's the ownership can be clearly specified by giving preference to the move semantics and avoiding shared ownership. This then allows to design APIs which enforce owner change and eliminate mistakes earlier (potentially in compile time, requiring an std::move call).

Reply:
I'd prefer that we stay with the functional approach and instead require a CreateProgram to have a Device parameter whereby there is no "move-to-device" at all.   Ideally we bound the program to device for its lifetime and if the user wants to create a new program they would have a different device handle.

Reply:
> require a CreateProgram to have a Device parameter

Is that actually desirable?

I think of
* Program can exist even without a device. 
* We could compile it without a device. 
* We could store compiled binaries somewhere.
* It’s okay to never send a program to any device

Reply:
Understood.  @cfjchu ,   I'd like your thoughts here as well.  Owning a DeviceHandle or ProgramHandle does not imply ownership of the underlying resource.   Having a move only handle sounds odd.   
The nice thing about the functional approach with handles to opaque resources is that we can add additional functions to handle specific cases as necessary.

Reply:
We synced offline but just to add here:

> Ideally we bound the program to device for its lifetime and if the user wants to create a new program they would have a different device handle.

We discussed the program does not need to bind its lifetime to device.

> I think of
> 
> Program can exist even without a device.
> We could compile it without a device.
> We could store compiled binaries somewhere.
> It’s okay to never send a program to any device

> We also want the ability to send the same program to multiple devices.

Agreed. Today, Device and Program are already coupled but with the proposed changes, user will just interface with a lightweight opaque handle. In the private impl, we can freely clean up the ownership semantics to decouple this relationship.

Reply:
All, we are currently seeing the ProgramHandle as being "move only".   That means that we move the program to a device or move it to some other device.